File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/PmdSecret.swift
//  Copyright © 2023 Polar. All rights reserved.

import Foundation
import CryptoKit
import CommonCrypto

public struct PmdSecret {
    let strategy: SecurityStrategy
    let key: Data
    private let keySymmetric: SymmetricKey?
    
    init(strategy: SecurityStrategy, key: Data) throws {
        switch(strategy) {
        case .none:
            guard key.isEmpty else {
                throw BleGattException.gattSecurityError(description: "key shall be empty for \(SecurityStrategy.none), key size was \(key.count)")
            }
            self.keySymmetric = nil
        case .xor:
            guard !key.isEmpty else {
                throw BleGattException.gattSecurityError(description: "key shall not be empty for \(SecurityStrategy.xor), key size was \(key.count)")
            }
            self.keySymmetric = nil
        case .aes128:
            guard key.count == 16 else {
                throw BleGattException.gattSecurityError(description: "key must be size of 16 bytes for \(SecurityStrategy.aes128), key size was \(key.count)")
            }
            self.keySymmetric = SymmetricKey(data: key)
            
        case .aes256:
            guard key.count == 32 else {
                throw BleGattException.gattSecurityError(description: "key must be size of 32 bytes for \(SecurityStrategy.aes256), key size was \(key.count)")
            }
            self.keySymmetric = SymmetricKey(data: key)
        }
        self.strategy = strategy
        self.key = key
    }
    
    func serializeToPmdSettings() -> Data {
        switch(self.strategy) {
        case .none:
            let securitySetting = PmdSetting.PmdSettingType.security.rawValue
            let securityStrategyByte = SecurityStrategy.none.rawValue
            let length:UInt8 = 1
            return Data([securitySetting, length, securityStrategyByte])
        case .xor:
            let securitySetting = PmdSetting.PmdSettingType.security.rawValue
            let securityStrategyByte = SecurityStrategy.xor.rawValue
            let keyBytes = key
            let length:UInt8 = 1
            return [securitySetting, length, securityStrategyByte] + keyBytes
        case .aes128:
            let securitySetting = PmdSetting.PmdSettingType.security.rawValue
            let securityStrategyByte = SecurityStrategy.aes128.rawValue
            let keyBytes = key
            let length:UInt8 = 1
            return [securitySetting, length, securityStrategyByte] + keyBytes
        case .aes256:
            let securitySetting = PmdSetting.PmdSettingType.security.rawValue
            let securityStrategyByte = SecurityStrategy.aes256.rawValue
            let keyBytes = key
            let length:UInt8 = 1
            return [securitySetting, length, securityStrategyByte] + keyBytes
        }
    }
    
    func decryptArray(cipherArray: Data) throws -> Data {
        switch(self.strategy) {
        case .none:
            return cipherArray
        case .xor:
            return Data(cipherArray.map { $0^key.first! })
        case .aes128, .aes256:
            return try decryptAES(chipperData: cipherArray)
        }
    }
    
    private func decryptAES(chipperData: Data) throws -> Data {
        guard chipperData.count % 16 == 0 else {
            throw BleGattException.gattSecurityError(description: "AES decryption failed. Chipper data is not dividable by 16, chipper size was \(chipperData.count)")
        }
        
        let cryptoLength = chipperData.count + key.count
        var cryptoData = Data(count: cryptoLength)
        var bytesLength = Int(0)
        
        let status = cryptoData.withUnsafeMutableBytes { cryptoBytes in
            chipperData.withUnsafeBytes { dataBytes in
                         key.withUnsafeBytes { keyBytes in
                    CCCrypt(UInt32(kCCDecrypt), CCAlgorithm(kCCAlgorithmAES), CCOptions(kCCOptionECBMode), keyBytes.baseAddress, key.count, nil, dataBytes.baseAddress, chipperData.count, cryptoBytes.baseAddress, cryptoLength, &bytesLength)
                }
            }
        }
        
        guard status == kCCSuccess else {
            debugPrint("Error: Failed to crypt data. Status \(status)")
            throw BleGattException.gattSecurityError(description: "AES decryption failed. Failed to error \(status)")
        }
        
        cryptoData.removeSubrange(bytesLength..<cryptoData.count)
        return cryptoData
    }
    
    enum SecurityStrategy: UInt8 {
        case none = 0
        case xor = 1
        case aes128 = 2
        case aes256 = 3
        
        static func fromByte(strategyByte: UInt8) throws -> SecurityStrategy {
            switch(strategyByte) {
            case SecurityStrategy.none.rawValue:
                return SecurityStrategy.none
            case SecurityStrategy.xor.rawValue:
                return SecurityStrategy.xor
            case SecurityStrategy.aes128.rawValue:
                return SecurityStrategy.aes128
            case SecurityStrategy.aes256.rawValue:
                return SecurityStrategy.aes256
            default :
                throw BleGattException.gattSecurityError(description: "Cannot decide security strategy from byte \(String(format:"%02X", strategyByte))")
            }
        }
    }
}

File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/PmdSecret.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd

import com.polar.androidcommunications.api.ble.exceptions.SecurityError
import javax.crypto.Cipher
import javax.crypto.spec.SecretKeySpec
import kotlin.experimental.xor

class PmdSecret(val strategy: SecurityStrategy, val key: ByteArray) {
    constructor(strategy: SecurityStrategy, key: SecretKeySpec) : this(strategy, key.encoded)

    init {
        when (strategy) {
            SecurityStrategy.NONE -> require(key.isEmpty()) { "key shall be empty for ${SecurityStrategy.NONE}, key size was ${key.size}" }
            SecurityStrategy.XOR -> require(key.isNotEmpty()) { "key shall not be empty for ${SecurityStrategy.XOR}, key size was ${key.size}" }
            SecurityStrategy.AES128 -> require(key.size == 16) { "key must be size of 16 bytes for ${SecurityStrategy.AES128}, key size was ${key.size}" }
            SecurityStrategy.AES256 -> require(key.size == 32) { "key must be size of 32 bytes for ${SecurityStrategy.AES256}, key size was ${key.size}" }
        }
    }

    fun serializeToPmdSettings(): ByteArray {
        when (strategy) {
            SecurityStrategy.NONE -> {
                val securitySetting = PmdSetting.PmdSettingType.SECURITY.numVal.toByte()
                val securityStrategyByte = SecurityStrategy.NONE.numVal.toByte()
                val length = 1.toByte()
                return byteArrayOf(securitySetting, length, securityStrategyByte)
            }
            SecurityStrategy.XOR -> {
                val securitySetting = PmdSetting.PmdSettingType.SECURITY.numVal.toByte()
                val securityStrategyByte = SecurityStrategy.XOR.numVal.toByte()
                val keyBytes = key
                val length = 1.toByte()
                return byteArrayOf(securitySetting, length, securityStrategyByte) + keyBytes
            }
            SecurityStrategy.AES128 -> {
                val securitySetting = PmdSetting.PmdSettingType.SECURITY.numVal.toByte()
                val securityStrategyByte = SecurityStrategy.AES128.numVal.toByte()
                val keyBytes = key
                val length = 1.toByte()
                return byteArrayOf(securitySetting, length, securityStrategyByte) + keyBytes
            }
            SecurityStrategy.AES256 -> {
                val securitySetting = PmdSetting.PmdSettingType.SECURITY.numVal.toByte()
                val securityStrategyByte = SecurityStrategy.AES256.numVal.toByte()
                val keyBytes = key
                val length = 1.toByte()
                return byteArrayOf(securitySetting, length, securityStrategyByte) + keyBytes
            }
        }
    }

    fun decryptArray(cipherArray: ByteArray): ByteArray {
        when (this.strategy) {
            SecurityStrategy.AES128 -> {
                val key = SecretKeySpec(this.key, "AES")
                val cipher = Cipher.getInstance("AES_128/ECB/NoPadding")
                cipher.init(Cipher.DECRYPT_MODE, key)
                return cipher.doFinal(cipherArray)
            }
            SecurityStrategy.AES256 -> {
                val key = SecretKeySpec(this.key, "AES")
                val cipher = Cipher.getInstance("AES_256/ECB/NoPadding")
                cipher.init(Cipher.DECRYPT_MODE, key)
                return cipher.doFinal(cipherArray)
            }
            SecurityStrategy.XOR -> {
                return cipherArray.map { it xor this.key.first() }.toByteArray()
            }
            SecurityStrategy.NONE -> {
                return cipherArray
            }
        }
    }

    enum class SecurityStrategy(val numVal: UByte) {
        NONE(0x00u),
        XOR(0x01u),
        AES128(0x02u),
        AES256(0x03u);

        companion object {
            fun fromByte(strategyByte: UByte): SecurityStrategy {
                return when {
                    (strategyByte == NONE.numVal) -> NONE
                    (strategyByte == XOR.numVal) -> XOR
                    (strategyByte == AES128.numVal) -> AES128
                    (strategyByte == AES256.numVal) -> AES256
                    else -> throw SecurityError.SecurityStrategyUnknown("Cannot decide security strategy from byte  ${"0x%x".format(strategyByte)}")
                }
            }
        }
    }
}